### ISSUE TITLE
appclient not workd with jboss-app-client-ear module

### ISSUE THREAD

**Issue by dmitrykolesnikovich** [jboss-developer/jboss-eap-quickstarts#1989]
There is an log for ClassNotFoundException:

D:\workspace\jboss-eap-quickstarts\app-client\ear\target>appclient.bat jboss-app-client.ear#simpleClient.jar Hello from command line
java.lang.ClassNotFoundException: jboss-app-client.ear#simpleClient.jar
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at org.glassfish.appclient.client.AppClientFacade.createContainerForClassName(AppClientFacade.java:482)
        at org.glassfish.appclient.client.AppClientFacade.createContainer(AppClientFacade.java:431)
        at org.glassfish.appclient.client.AppClientFacade.prepareACC(AppClientFacade.java:270)
        at org.glassfish.appclient.client.acc.agent.AppClientContainerAgent.premain(AppClientContainerAgent.java:83)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
        at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)

**Comment by sgilda** [comment]
@dmitrykolesnikovich : 

From your trace, it looks like you are running the app-client.bat command from the `D:\workspace\jboss-eap-quickstarts\app-client/ear/target` directory instead of from the root of the quickstart directory `D:\workspace\jboss-eap-quickstarts\app-client` directory.

1. Be sure to navigate to the root directory of this quickstart: 

        cd D:\workspace\jboss-eap-quickstarts\app-client

2. Then build the quickstart using this command:  

        mvn clean install wildfly:deploy

3. Open the D:\workspace\jboss-eap-quickstarts\app-client\ear\jboss-app-client.ear file and verify that it contains the following structure:

    ├── ejb.jar
    ├── META-INF
    │   ├── application.xml
    │   └── MANIFEST.MF
    └── simpleClient.jar

4. Still in the cd D:\workspace\jboss-eap-quickstarts\app-client\ directory, type the following command, replacing EAP7_HOME with the location of your server installation.

    EAP7_HOME/bin/appclient.batear\target\jboss-app-client.ear#simpleClient.jar Hello from command line

5. Finally, it appears your quickstart project is located within the IDE workspace. You need to move it to a location outside of the IDE workspace. See the documentation here for an explanation: https://access.redhat.com/documentation/en/red-hat-jboss-enterprise-application-platform/version-7.0/getting-started-guide/#run_the_quickstarts_in_red_hat_jboss_developer_studio

Let me know if these steps resolve your issue.

**Comment by dmitrykolesnikovich** [comment]
Still not working. Here is video https://www.youtube.com/watch?v=FGyW_Hqd83g

**Comment by sgilda** [comment]
From the video, it still appears that you including the path to your JBoss EAP server installation when you run the app-client.bat command. The app-client.bat file is located in the server installation bin directory, not the root of the app-client quickstart.

Try this command from the root of the app-client quickstart directory and let me know if it works: 

      c:\Users\kolesnikovich\EAP-7.0.0\bin\app-client.bat ear\target\jboss-app-client.ear#simpleClient.jar Hello from command line

**Comment by dmitrykolesnikovich** [comment]
Now it works. It gives me a lot info in logs beside **actual system out print**. Is that expected behaviour? By the way, why appclient not working (I included EAP/bin path in PATH windows variable).

D:\workspace\jboss-eap-quickstarts\app-client>c:\Users\kolesnikovich\EAP-7.0.0\bin\appclient.bat ear\target\jboss-app-client.ear#simpleClient.jar Hello from command line
Calling "c:\Users\kolesnikovich\EAP-7.0.0\bin\appclient.conf.bat"
18:55:55,226 INFO  [org.jboss.modules] (main) JBoss Modules version 1.5.1.Final-redhat-1
18:55:55,693 INFO  [org.jboss.msc] (main) JBoss MSC version 1.2.6.Final-redhat-1
18:55:55,875 INFO  [org.jboss.as] (MSC service thread 1-3) WFLYSRV0049: JBoss EAP 7.0.0.GA (WildFly Core 2.1.2.Final-redhat-1) starting
18:55:58,219 INFO  [org.jboss.as.connector.subsystems.datasources] (ServerService Thread Pool -- 16) WFLYJCA0004: Deploying JDBC-compliant driver class org.h2.Driver (version 1.3)
18:55:58,308 INFO  [org.jboss.as.security] (ServerService Thread Pool -- 23) WFLYSEC0002: Activating Security Subsystem
18:55:58,303 INFO  [org.jboss.as.webservices] (ServerService Thread Pool -- 25) WFLYWS0002: Activating WebServices Extension
18:55:58,333 WARN  [org.jboss.as.txn] (ServerService Thread Pool -- 24) WFLYTX0013: Node identifier property is set to the default value. Please make sure it is unique.
18:55:58,340 INFO  [org.jboss.as.naming] (ServerService Thread Pool -- 22) WFLYNAM0001: Activating Naming Subsystem
18:55:58,353 INFO  [org.wildfly.iiop.openjdk] (ServerService Thread Pool -- 19) WFLYIIOP0001: Activating IIOP Subsystem
18:55:58,426 INFO  [org.jboss.as.security] (MSC service thread 1-4) WFLYSEC0001: Current PicketBox version=4.9.6.Final-redhat-1
18:55:58,443 INFO  [org.jboss.as.naming] (MSC service thread 1-3) WFLYNAM0003: Starting Naming Service
18:55:59,442 INFO  [org.jboss.as.connector] (MSC service thread 1-4) WFLYJCA0009: Starting JCA Subsystem (WildFly/IronJacamar 1.3.3.Final-redhat-1)
18:55:59,602 INFO  [org.jboss.as.connector.deployers.jdbc] (MSC service thread 1-4) WFLYJCA0018: Started Driver servicewith driver-name = h2
18:55:59,665 INFO  [org.jboss.ws.common.management] (MSC service thread 1-1) JBWS022052: Starting JBossWS 5.1.3.SP1-redhat-1 (Apache CXF 3.1.4.redhat-1)
18:56:00,433 INFO  [org.wildfly.iiop.openjdk] (MSC service thread 1-4) WFLYIIOP0009: CORBA ORB Service started
18:56:00,539 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-1) WFLYJCA0001: Bound data source [java:jboss/datasources/ExampleDS]
18:56:00,731 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-2) WFLYSRV0027: Starting deployment of "jboss-app-client.ear" (runtime-name: "jboss-app-client.ear")
18:56:00,973 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-2) WFLYSRV0207: Starting subdeployment (runtime-name: "ejb.jar")
18:56:00,974 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-2) WFLYSRV0207: Starting subdeployment (runtime-name: "simpleClient.jar")
18:56:01,226 INFO  [org.jboss.as.ejb3.deployment] (MSC service thread 1-3) WFLYEJB0473: JNDI bindings for session bean named 'StatelessSessionBean' in deployment unit 'subdeployment "ejb.jar" of deployment "jboss-app-client.ear"' are as follows:

        java:global/jboss-app-client/ejb/StatelessSessionBean!org.jboss.as.quickstarts.appclient.server.ejb.StatelessSes
sion
        java:app/ejb/StatelessSessionBean!org.jboss.as.quickstarts.appclient.server.ejb.StatelessSession
        java:module/StatelessSessionBean!org.jboss.as.quickstarts.appclient.server.ejb.StatelessSession
        java:jboss/exported/jboss-app-client/ejb/StatelessSessionBean!org.jboss.as.quickstarts.appclient.server.ejb.Stat
elessSession
        java:global/jboss-app-client/ejb/StatelessSessionBean
        java:app/ejb/StatelessSessionBean
        java:module/StatelessSessionBean

18:56:01,483 INFO  [org.jboss.as.server] (Thread-34) WFLYSRV0010: Deployed "jboss-app-client.ear" (runtime-name : "jboss-app-client.ear")
18:56:01,491 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: JBoss EAP 7.0.0.GA (WildFly Core 2.1.2.Final-redhat-1) started in 5855ms - Started 227 of 230 services (18 services are lazy, passive or on-demand)
18:56:01,528 INFO  [org.jboss.ejb.client] (Thread-41) JBoss EJB Client version 2.1.4.Final-redhat-1
**18:56:01,552 INFO  [org.jboss.as.quickstarts.appclient.acc.client.Main] (Thread-41) Main started with arguments
18:56:01,553 INFO  [org.jboss.as.quickstarts.appclient.acc.client.Main] (Thread-41)             [Hello, from, command, line]**
18:56:01,569 INFO  [org.xnio] (Thread-41) XNIO version 3.3.6.Final-redhat-1
18:56:01,598 INFO  [org.xnio.nio] (Thread-41) XNIO NIO Implementation Version 3.3.6.Final-redhat-1
18:56:01,655 INFO  [org.jboss.remoting] (Thread-41) JBoss Remoting version 4.0.18.Final-redhat-1
18:56:01,927 INFO  [org.jboss.ejb.client.remoting] (Remoting "endpoint" task-5) EJBCLIENT000017: Received server version 2 and marshalling strategies [river]
18:56:01,940 INFO  [org.jboss.ejb.client.remoting] (Thread-41) EJBCLIENT000013: Successful version handshake completed for receiver context EJBReceiverContext{clientContext=org.jboss.ejb.client.EJBClientContext@aa11654, receiver=Remoting connection EJB receiver [connection=Remoting connection <7956c718>,channel=jboss.ejb,nodename=kolesnikovichpc]} on channel
 Channel ID c27a8ce1 (outbound) of Remoting connection 6be033b0 to localhost/127.0.0.1:8080
18:56:02,031 INFO  [org.jboss.as.quickstarts.appclient.acc.client.interceptor.ClientInterceptor] (Thread-41) Adding jboss.node.name (kolesnikovichpc) to the invocation context
18:56:02,165 INFO  [org.jboss.as.quickstarts.appclient.acc.client.Main] (Thread-41) Hello from StatelessSessionBean@kolesnikovichpc
18:56:02,171 INFO  [org.jboss.as.quickstarts.appclient.acc.client.interceptor.ClientInterceptor] (Thread-41) Adding jboss.node.name (kolesnikovichpc) to the invocation context
18:56:02,245 INFO  [org.jboss.ejb.client.remoting] (Remoting "endpoint" task-9) EJBCLIENT000016: Channel Channel ID c27a8ce1 (outbound) of Remoting connection 6be033b0 to localhost/127.0.0.1:8080 can no longer process messages
18:56:02,367 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-2) WFLYJCA0010: Unbound data source [java:jboss/datasources/ExampleDS]
18:56:02,411 INFO  [org.hibernate.validator.internal.util.Version] (MSC service thread 1-1) HV000001: Hibernate Validator 5.2.4.Final-redhat-1
18:56:02,512 INFO  [org.jboss.as.connector.deployers.jdbc] (MSC service thread 1-1) WFLYJCA0019: Stopped Driver servicewith driver-name = h2
18:56:02,570 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-2) WFLYSRV0208: Stopped subdeployment (runtime-name: ejb.jar) in 256ms
18:56:02,594 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-1) WFLYSRV0208: Stopped subdeployment (runtime-name: simpleClient.jar) in 280ms
18:56:02,608 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-1) WFLYSRV0028: Stopped deployment jboss-app-client.ear (runtime-name: jboss-app-client.ear) in 294ms
18:56:02,653 INFO  [org.jboss.as] (MSC service thread 1-4) WFLYSRV0050: JBoss EAP 7.0.0.GA (WildFly Core 2.1.2.Final-redhat-1) stopped in 338ms

**Comment by sgilda** [comment]
@dmitrykolesnikovich : 

Yes, you should see the above "INFO" level output from your Maven console and your server logs. Those are informational messages so you know exactly what is happening. You can configure your application and server logging to see more or less information by following the instructions in this chapter of the Configuration Guide: https://access.redhat.com/documentation/en/red-hat-jboss-enterprise-application-platform/version-7.0/configuration-guide/#logging_with_jboss_eap

I think your environment variable for the path is working. You just ran it from the wrong directory. 

You ran it from the app-client\ear\target directory as follows:

D:\workspace\jboss-eap-quickstarts\app-client\ear\target>appclient.bat jboss-app-client.ear#simpleClient.jar Hello from command line

Try running it from the root of the app-client directory as I suggested in the first reply above:

D:\workspace\jboss-eap-quickstarts\app-client>appclient.bat ear\target\/jboss-app-client.ear#simpleClient.jar Hello from command line

**Comment by dmitrykolesnikovich** [comment]
Nope, it does not work from root directory. That's why I ask.

D:\workspace\jboss-eap-quickstarts\app-client>appclient.bat ear\target\/jboss-app-client.ear#simpleClient.jar Hello from command line
java.lang.ClassNotFoundException: ear\target\/jboss-app-client.ear#simpleClient.jar
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at org.glassfish.appclient.client.AppClientFacade.createContainerForClassName(AppClientFacade.java:482)
        at org.glassfish.appclient.client.AppClientFacade.createContainer(AppClientFacade.java:431)
        at org.glassfish.appclient.client.AppClientFacade.prepareACC(AppClientFacade.java:270)
        at org.glassfish.appclient.client.acc.agent.AppClientContainerAgent.premain(AppClientContainerAgent.java:83)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
        at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)

D:\workspace\jboss-eap-quickstarts\app-client>c:\Users\kolesnikovich\EAP-7.0.0\bin\appclient.bat ear\target\/jboss-app-client.ear#simpleClient.jar Hello from command line
Calling "c:\Users\kolesnikovich\EAP-7.0.0\bin\appclient.conf.bat"
22:29:29,353 INFO  [org.jboss.modules] (main) JBoss Modules version 1.5.1.Final-redhat-1
...

May be the reason is that script can not find "c:\Users\kolesnikovich\EAP-7.0.0\bin\appclient.conf.bat" if I do not run it with full path.

**Comment by sgilda** [comment]
@dmitrykolesnikovich : I was able to find a Windows machine to test this out and it worked fine for me.

1. I added `jboss-eap-7.0\bin` to my path.
2. I added the users to the server.
3. I started the server in one command prompt.
4. I opened another command prompt and navigated to `jboss-eap-7.0.0.GA-quickstarts\app-client`
5. I built and deployed the quickstart using `mvn clean install wildfly:deploy`
6. From the same directory I ran this command `appclient.bat ear\target\jboss-app-client.ear#simpleClient.jar Hello from command line

It ran with no error and I got the expected results in the console and in the server.

Looking at what you pasted above, I wonder if your problem is the extra forward slash before the `jboss-app-client`?

Your command is this:

appclient.bat ear\target\/jboss-app-client.ear#simpleClient.jar Hello from command line
 
Try this:

appclient.bat ear\target\jboss-app-client.ear#simpleClient.jar Hello from command line
